app/src/main/java/com/zeapo/pwdstore/autofill/oreo/Api30AutofillResponseBuilder.kt

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright © 2014-2020 The Android Password Store Authors. All Rights Reserved.
+ * SPDX-License-Identifier: GPL-3.0-only
+ */
+
+package com.zeapo.pwdstore.autofill.oreo
+
+import android.content.Context
+import android.content.IntentSender
+import android.os.Build
+import android.service.autofill.Dataset
+import android.service.autofill.FillCallback
+import android.service.autofill.FillResponse
+import android.service.autofill.SaveInfo
+import android.view.inputmethod.InlineSuggestionsRequest
+import android.widget.inline.InlinePresentationSpec
+import androidx.annotation.RequiresApi
+import com.github.ajalt.timberkt.e
+import com.github.androidpasswordstore.autofillparser.AutofillAction
+import com.github.androidpasswordstore.autofillparser.FillableForm
+import com.github.androidpasswordstore.autofillparser.fillWith
+import com.github.michaelbull.result.fold
+import com.zeapo.pwdstore.autofill.oreo.ui.AutofillDecryptActivity
+import com.zeapo.pwdstore.autofill.oreo.ui.AutofillFilterView
+import com.zeapo.pwdstore.autofill.oreo.ui.AutofillPublisherChangedActivity
+import com.zeapo.pwdstore.autofill.oreo.ui.AutofillSaveActivity
+import com.zeapo.pwdstore.autofill.oreo.ui.AutofillSmsActivity
+import java.io.File
+
+/**
+ * Implements [AutofillResponseBuilder]'s methods for API 30 and above
+ */
+@RequiresApi(Build.VERSION_CODES.R)
+class Api30AutofillResponseBuilder(form: FillableForm) {
+
+    private val formOrigin = form.formOrigin
+    private val scenario = form.scenario
+    private val ignoredIds = form.ignoredIds
+    private val saveFlags = form.saveFlags
+    private val clientState = form.toClientState()
+
+    // We do not offer save when the only relevant field is a username field or there is no field.
+    private val scenarioSupportsSave = scenario.hasPasswordFieldsToSave
+    private val canBeSaved = saveFlags != null && scenarioSupportsSave
+
+    private fun makeIntentDataset(
+        context: Context,
+        action: AutofillAction,
+        intentSender: IntentSender,
+        metadata: DatasetMetadata,
+        imeSpec: InlinePresentationSpec?,
+    ): Dataset {
+        return Dataset.Builder(makeRemoteView(context, metadata)).run {
+            fillWith(scenario, action, credentials = null)
+            setAuthentication(intentSender)
+            if (imeSpec != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
```

Comment:
Could we get rid of the checks here now that this is R+ only?

Reply:
Yep, removed.